### PR TITLE
Introduce introspection operations

### PR DESCRIPTION
**Pull request by sanket1729**
- Equality of values, assets, spks(possibly confidential)
- Testing where an asset/value is explicit
- Checking the current executing input index

### CODE REVIEW

**Comment by sanket1729** [comment]
The last commit adds a detailed spec for extensions that should make it easy to review and share with others interested in creating covenants on liquid/elements

**Comment by apoelstra** [comment]
Wow! This is a really cool PR. I'll ACK it when my local scripts finish -- my comments are just nits and/or could be dealt with in future PRs.

**Review by apoelstra — APPROVED** [review]
ACK 3dce073fd80041e9afa8d9ffdcc80ac2f26a8791

File: src/extensions/introspect_ops.rs [hunk 1]
```diff
@@ -0,0 +1,1174 @@
...
+            base: Base::B,
+            input: Input::Zero, // No input from stack
+            dissatisfiable: true,
+            unit: true,
+        }
+    }
+
+    fn mall_prop(&self) -> Malleability {
+        Malleability {
+            dissat: Dissat::Unknown, // many dissatisfactions possible
```
[the remark this answers is not in the document]
Comment (sanket1729) [reply src/extensions/introspect_ops.rs:418]: The transaction itself can be changed. For example, if someone re-blinds the first output of tx to some other value. The fragment `asset_eq()` can have multiple dissatisfactions.

File: src/extensions/introspect_ops.rs [hunk 2]
```diff
@@ -0,0 +1,1174 @@
...
+            safe: false,             // Unsafe as a top fragment
+            non_malleable: true,
+        }
+    }
+
+    fn extra_prop(&self) -> ExtData {
+        ExtData {
+            pk_cost: self.script_size(), // 1 opcodes, 1 key push, msg, 1 msg push
+            has_free_verify: false,
```
[the remark this answers is not in the document]
Comment (sanket1729) [reply src/extensions/introspect_ops.rs:427]: I think this should be possible to do. The lexer should handle both `EQUAL VERIFY` and `EQUALVERIFY` correctly.

[the remark this answers is not in the document]
Comment (sanket1729) [reply src/extensions/introspect_ops.rs:427]: I initially had `NIP EQUALVERIFY` as the fragment. But it does not interact nicely with our current parser as it tries to parse a threshW.

File: src/extensions/introspect_ops.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/extensions/introspect_ops.rs:418]: Ok, cool. This feels like an extension of what `Dissat::Unkwown` means but it makes sense.

File: src/extensions/introspect_ops.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/extensions/introspect_ops.rs:427]: Frustrating. I think for now we'll just live with it. We can double-back and try to hyperoptimize things when we're closer to having production deployments.

**Pull request merged**